cookbooks/cdo-varnish/test/integration/cloudfront/minitest/test_default.rb

```diff
@@ -0,0 +1,315 @@
+  `curl -X #{method} -s #{cookie_string} #{header_string} -i #{url}`.tap{assert_equal 0, $?.exitstatus}
+end
+
+def request(url, headers={}, cookies={})
+  _request("#{LOCALHOST}:#{DASHBOARD_PORT}#{url}", headers, cookies)
+end
+
+def proxy_request(url, headers={}, cookies={}, method='GET')
+  headers.merge!(host: 'cdo.ngrok.io')
+  headers.merge!('X-Forwarded-Proto' => 'https')
+  _request("#{CLOUDFRONT_URL}#{url}", headers, cookies, method)
+end
+
+def code(response)
+  /HTTP\/1\.1 (\d+)/.match(response.lines.to_a.first.strip)[1].to_i
+end
+
+def assert_ok(response)
+  assert_equal 200, code(response)
+end
+def assert_cache(response, hit)
+  assert_equal hit ? 'Hit' : 'Miss', /X-Cache: (\w+) from cloudfront/.match(response)[1]
+end
+def assert_miss(response)
+  assert_cache response, false
+end
+def assert_hit(response)
+  assert_cache response, true
+end
+
+def last_line(response)
+  response.lines.to_a.last.strip
+end
+
+require 'securerandom'
+def get_url(prefix='x', suffix='')
+  id = SecureRandom.uuid
+  '/' + prefix.to_s + '/' + id + '/' + suffix.to_s
+end
+
+describe 'http' do
+  it 'handles a simple request' do
+    url = get_url 1
+    text = 'Hello World!'
+    mock_url(url, text)
+    response = request url
+    assert_ok response
+    assert_equal text, last_line(response)
+  end
+
+  it 'caches a simple request' do
+    url = get_url 2
+    text = 'Hello World 2!'
+    mock_url(url, text)
+
+    response = proxy_request url
+    assert_ok response
+    assert_miss response
+
+    response = proxy_request url
+    assert_ok response
+    assert_hit response
+
+    # Verify only one request hit the origin server
+    assert_equal 1, count_requests(url)
+  end
+
+  it 'redirects HTTP to HTTPS' do
+    response = _request "#{LOCALHOST}:#{VARNISH_PORT}/https"
+    assert_equal 301, code(response)
+    assert_equal "https://#{LOCALHOST}/https", /Location: ([^\s]+)/.match(response)[1]
+  end
+
+  it 'Normalizes Accept-Language' do
+    url = get_url 3
+    text_en = 'Hello World!'
+    text_fr = 'Bonjour le Monde!'
+    mock_url(url, text_en, {'X-Varnish-Accept-Language' => 'en'}, {vary: 'X-Varnish-Accept-Language'})
+    mock_url(url, text_fr, {'X-Varnish-Accept-Language' => 'fr'}, {vary: 'X-Varnish-Accept-Language'})
+    en = {'Accept-Language' => 'en'}
+    response = proxy_request url, en
+    assert_miss response
+    assert_equal text_en, last_line(response)
+
+    # Ensure that Vary response header is de-normalized
+    assert_nil /X-Varnish-Accept-Language/.match(response)
+    refute_nil /Accept-Language/.match(response)
+    assert_hit proxy_request url, en
+
+    # Ensure that language is separately cached
+    fr = {'Accept-Language' => 'fr'}
+    response = proxy_request url, fr
+    assert_miss response
+    assert_equal text_fr, last_line(response)
+    assert_hit proxy_request url, fr
+
+    # Language-header normalization not implemented in CloudFront.
```

Comment:
That sounds fine.  Or you could have a generic cache method that is happy
if either varnish or cloudfront gets a hit.  (That way, if we make
CloudFront smarter down the road, the test doesn't break needlessly, since
we're still getting the desired hit.)

On Thu, Oct 8, 2015 at 4:53 PM, Will Jordan <notifications@github.com>
wrote:

> In
> cookbooks/cdo-varnish/test/integration/cloudfront/minitest/test_default.rb
> <https://github.com/code-dot-org/code-dot-org/pull/4440#discussion_r41584705>
> :
>
> > +    assert_miss response
> > +    assert_equal text_en, last_line(response)
> > +
> > +    # Ensure that Vary response header is de-normalized
> > +    assert_nil /X-Varnish-Accept-Language/.match(response)
> > +    refute_nil /Accept-Language/.match(response)
> > +    assert_hit proxy_request url, en
> > +
> > +    # Ensure that language is separately cached
> > +    fr = {'Accept-Language' => 'fr'}
> > +    response = proxy_request url, fr
> > +    assert_miss response
> > +    assert_equal text_fr, last_line(response)
> > +    assert_hit proxy_request url, fr
> > +
> > +    # Language-header normalization not implemented in CloudFront.
>
> This test will succeed on a CloudFront + Varnish request only if we check
> the Varnish X-Varnish-Cache header for Hit, rather than the CloudFront
> X-Cache one. To distinguish this test as specifically checking the
> Varnish cache header, I'll add a new method assert_varnish_cache.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/4440/files#r41584705>.
>